### DISCUSSION TITLE
Custom button to load in AR doesn't take into account dynamic changes to model when using Lazy loading

### DISCUSSION THREAD

**Discussion by da3ba** [google/model-viewer#4475] [Q&A] [answered]
I'm using the lazy loading feature in addition to a custom button to enable AR, but when clicking the Enable AR button it doesn't take under consideration the new texture changed dynamically for the model. I want to be able to see the new texture once I click the enable AR button without the need to activate the button-load.

```
 <model-viewer id="lazy-load" src="{{ source_glb }}" alt="A 3D model" reveal="manual" style="height: 336px; width: 100%; padding-bottom: 2rem;" {%- if block.settings.wall_mode_enabled -%} ar ar-placement="wall" {%- endif -%}  camera-controls touch-action="pan-y" ar ar-scale="fixed" {%- if block.settings.auto_rotate_enabled -%} auto-rotate {%- endif -%}>
      
       <div id="lazy-load-poster" style="background-image: url('{{ block.settings.bg-image.src | img_url: 'master' }}')" slot="poster">
          <div id="button-load">Load 3D Model</div>
        </div>
        <button slot="ar-button" class="button ar-button-load" style="position: absolute;">
          Activate AR
        </button>
      </model-viewer>

// Define a flag to track whether the model has been loaded
let modelLoaded = false;

// Function to load the model, update texture and scale, and activate AR mode
async function loadModelAndUpdateAR(newTextureUrl) {
  if (!modelLoaded) {
    // Wait for the model to load
    await new Promise((resolve) => {
      modelViewer.addEventListener('load', () => {
        modelLoaded = true;

        // Find the material you want to change (adjust the index as needed)
        const materialIndex = 0; // Change this to the correct material index
        const material = modelViewer.model.materials[materialIndex];

        // Create a new texture
        const texture = await modelViewer.createTexture(newTextureUrl);

        // Set the new texture
        material.pbrMetallicRoughness['baseColorTexture'].setTexture(texture);

        // Rotate the texture 180 degrees (flip vertically)
        material.pbrMetallicRoughness['baseColorTexture'].texture.uvTransform = [1, -1, 1, 1, -1, 1, 0, 1];

        modelViewer.updateFraming();
        resolve();
      });
    });
  }

  // Activate AR mode
  modelViewer.activateAR();
}

// Debugging alert to check if the event listener is registered
console.log("Josh1");

// Listen for the AR button click event
document.querySelector('.ar-button-load').addEventListener('click', () => {
  document.querySelector('#lazy-load').dismissPoster();
  const newTextureUrl = "YOUR_TEXTURE_URL_HERE"; // Replace with the desired texture URL
  loadModelAndUpdateAR(newTextureUrl);
});


```

**Comment by elalish** [comment]
The primary information I need to answer a question about AR is what device you're using and what AR mode you're using. Do you get different behavior when using the default AR button?

**Reply by da3ba** [reply, marked as answer]
Thanks, I've been able to fix the issue by using a bit of gymnastics

document.querySelector('.ar-button-loadd').addEventListener('click', () => 
{
  // Listen for the custom event
  document.addEventListener(modelLoadedEventName, triggerARActivation);

  const modelViewer = document.querySelector('#lazy-load');

  if (!modelViewer) 
  {
    alert("UNDEFINED FOUND");
    return;
  }

  // Dismiss the poster immediately
  document.querySelector('#lazy-load').dismissPoster();
  //alert("CLICKED IT");

  // Check if the model is already loaded
  if (modelViewer.loaded) 
  {
    // If it's already loaded, trigger AR activation immediately
    triggerARActivation();
  } else {
    // If not, listen for the 'load' event and dispatch the custom event when it occurs
    modelViewer.addEventListener('load', () => {
      // Dispatch the custom event when the model is loaded
      document.dispatchEvent(new Event(modelLoadedEventName));
    });
  }
});,
